crates/bevy_render/src/render_phase/mod.rs

```diff
@@ -5,42 +5,75 @@ pub use draw::*;
 pub use draw_state::*;
 
 use bevy_ecs::prelude::{Component, Query};
+use std::cell::Cell;
+use thread_local::ThreadLocal;
 
 use copyless::VecHelper;
 
+pub struct RenderPhaseScope<'a, I: PhaseItem> {
+    phase_queue: &'a mut Vec<I>,
+}
+
+impl<'a, I: PhaseItem> RenderPhaseScope<'a, I> {
+    /// Adds a [`PhaseItem`] to this render phase.
+    #[inline]
+    pub fn add(&mut self, item: I) {
+        self.phase_queue.alloc().init(item);
+    }
+}
+
 /// A resource to collect and sort draw requests for specific [`PhaseItems`](PhaseItem).
 #[derive(Component)]
 pub struct RenderPhase<I: PhaseItem> {
-    pub items: Vec<I>,
+    items: ThreadLocal<Cell<Vec<I>>>,
+    pub sorted: Vec<I>,
 }
 
 impl<I: PhaseItem> Default for RenderPhase<I> {
     fn default() -> Self {
-        Self { items: Vec::new() }
+        Self {
+            items: ThreadLocal::new(),
+            sorted: Vec::new(),
+        }
     }
 }
 
 impl<I: PhaseItem> RenderPhase<I> {
-    /// Adds a [`PhaseItem`] to this render phase.
-    #[inline]
-    pub fn add(&mut self, item: I) {
-        self.items.alloc().init(item);
+    pub fn get(&self) -> &Cell<Vec<I>> {
+        self.items.get_or_default()
+    }
+
+    pub fn phase_scope(&self, f: impl FnOnce(RenderPhaseScope<'_, I>)) {
+        let store = self.get();
+        let mut phase_queue = store.take();
+        f(RenderPhaseScope {
+            phase_queue: &mut phase_queue,
+        });
+        store.set(phase_queue);
     }
 
     /// Sorts all of its [`PhaseItems`](PhaseItem).
     pub fn sort(&mut self) {
-        self.items.sort_by_key(|d| d.sort_key());
+        debug_assert!(self.sorted.is_empty());
+        for batch in self.items.iter_mut() {
+            self.sorted.append(batch.get_mut());
+        }
```

Comment:
You could preallocate the exact space needed in `self.sorted` to avoid potential reallocations.

```suggestion
        let length = self.items.iter_mut().map(|batch| batch.get_mut().len()).sum();
        self.sorted.reserve(length);
        for batch in self.items.iter_mut() {
            self.sorted.append(batch.get_mut());
        }
```

Reply:
I don't think there's any atomic operation involved here. The code is using `iter_mut` which requires exclusive access and thus can be sure this is the only thread accessing the `ThreadLocal`. 

In fact looking at the source code you can see it uses `*bucket.get_mut()` instead of `bucket.load(Ordering::Acquire)` and `*entry.present.get_mut()` instead of `entry.present.load(Ordering::Acquire)`

https://github.com/Amanieu/thread_local-rs/blob/00b0d9daefd7c8308b285dfd82be647ea1fb6b72/src/lib.rs#L383-L428